README.md
LTSM - Lightweight TSM API, Lustre TSM Copytool for Archiving and Retrieving Data.
==============

[![Tag Version](https://img.shields.io/github/tag/tstibor/ltsm.svg)](https://github.com/tstibor/ltsm/tags)
[![License](http://img.shields.io/:license-gpl2-blue.svg)](http://www.gnu.org/licenses/gpl-2.0.html)
[![Linux](https://github.com/tstibor/ltsm/actions/workflows/linux.yml/badge.svg)](https://github.com/tstibor/ltsm/actions)

This project consists of *four* parts:
1. Lightweight TSM API/library (called *tsmapi*) supporting operations (*archiving*, *retrieving*, *deleting*, *querying*).
2. Lustre TSM Copytool.
3. TSM console client.
4. Benchmark and test suite.

# Introduction into Lustre HSM
Lustre has since version 2.5 hierarchical storage management (HSM) capabilities, that is, data can be automatically
*archived* to low-cost storage media such as tape storage systems and seamlessly *retrieved* when accessing the data on Lustre clients. The Lustre HSM capabilities can be illustrated by means of state diagrams and the related Lustre commands
*lfs hsm_archive*, *lfs hsm_release*, *lfs hsm_restore* as depicted below.
![Statediagram](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/state.diagram.png)
The initial state is a *new file*. Once the file is archived, it
changes into state *archived* and a identical copy exists on the HSM storage. If the archived file is modified, then the state
changes to *dirty*. As a consequence, one has to re-archive the file to reobtain an identical copy on the Lustre side as well
as the HSM storage. If the state changes to *released*, then the bulk data of the file is moved to the HSM storage and only the file skeleton
is kept. If the file is accessed or the *lfs hsm_restore* command is executed, the file is restored and the state is changes back to *archived*.

# Lustre HSM Framework
The aim of the Lustre HSM framework is to provide an interface for seamlessly *archiving* data which is no
longer used, and place them into a long term retrievable archive storage (such as tape drives).
The meta information of the *archived* data such as the file name, user id, etc. is however still
present in the Lustre file system, whereas the bulk data is placed on the retrievable storage.
When the *released* data is accessed, the bulk data is seamlessly copied from the retrievable storage back to the Lustre
file system. In summary, the data is still available regardless of where it is actually stored.
For providing such a functionality, the HSM framework is embedded in the main Lustre components and explained
for the case of *retrieving* data in the Figure below.
![HSMArchitecture](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/hsm.architecture.png)
Triggered actions and data flows for retrieving data inside the Lustre HSM framework.
The participating nodes (depicted as dashed frames) are: Client, MDS, OSS, Copytool, Storage. The Lustre client opens a file which is already archived, thus the bulk data is not available on any OST. The MDT node receives the open call, allocate objects on the OST and notifies the HSM coordinator to copy in data for file indentifier (short FID) of the corresponding file. The copytool receives from the HSM coordinator the event to copyin data of FID and notifies the internal data mover for retrieving the data stored on the storage node and copy it over to a OST. Once the copy process is completed, the data mover notifies the HSM agent with a copydone message and the HSM agent finalizes the process with a HSM copydone message back to the MDT node.

# TSM Overview
Tivoli Storage Manager (now renamed to IBM Spectrum Protect) (TSM) is a client/server product from IBM
employed in heterogeneous distributed environments to *backup* and *archive* data.
The difference between a backup and an archive can be summarized as follows:
* Backup: A copy of the data is stored in the event the original becomes lost or damaged. Typically an incremental (forever) backup strategy is performed.
* Archive: Remove from an on-line system those data no longer in day to day use, and place them into a long term retrievable storage (such as tape drives).

The core capabilities and features of a TSM server are:
* Compression: Compress data stream seamless either on client or server side.
* Deduplication: Eliminating duplicate copies of repeating data.
* Collocation: Store and pack data of a client in few number of tapes as much as possible to reduce the number of media mounts and for minimizing tape drive movements.
* Storage hierarchies: Automatically move data from faster devices to slower devices based on characteristics such as file size or storage capacity.

The last issue is a crucial feature for the Lustre HSM framework, as it enables effectively archiving and retrieving data to fast devices, rather than being tied up to slow but large and cheap devices (see Fig. below).
![fig:storage.hierarchy](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/storage.hierarchy.png)
Storage hierarchy of a TSM server. Fast but limited space storage devices (high cost per GByte) are located at the
top of the pyramid, whereas slower but large space storage devices (low cost per GByte) are located at the bottom.
Fast storage devices act as a *cache* layer, where data in automatically migrated by the TSM server based on access pattern and
space occupation level. More details can be found in [IBM Tivoli Storage Manager Implementation Guide (pp. 257)](https://github.com/tstibor/ltsm.github.io/raw/master/doc/tsm/ibm_tivoli_storage_management_implementation_guide.pdf#page=287).

Denote r<sub>1</sub> the transfer rate to (fast) storage devices s<sub>1</sub> and r<sub>2</sub> the transfer rate to (slow) storage devices s<sub>2</sub>.
Make sure to set the migration threshold t from s<sub>1</sub> to s<sub>2</sub> to a proper value which suits your use case.
Example: Let r<sub>1</sub> = 800MB/sec, r<sub>2</sub> = 200MB/sec, s<sub>1</sub> = 150TB and migration threshold is set to t = 0.6.
If r<sub>1</sub> &#62; r<sub>2</sub>, then after t &#183; s<sub>1</sub> / (r<sub>1</sub> - r<sub>2</sub>) seconds, s<sub>1</sub> is 100% full. For that example it yields
41.66667 hours, when continuously writing into s<sub>1</sub> with rate r<sub>1</sub>.

# Data Organized on TSM Server
The TSM server is an object storage server and is developed for storing and retrieving named objects. Similar to Lustre, the TSM
server has two main storage areas, database and data storage. The database contains the metadata of the objects such as unique
identifiers, names and attributes, whereas the data storage contains the object data. Each object stored on the TSM server has a
name associated with it that is determined by the client. The object name is composed of:
* **fs**: File space name (mount point),
* **hl**: High level name (directory name),
* **ll**: Low level name (file name),

and must be specified to identify the object for operations query, archive, retrieve and delete. In addition to determined TSM
metadata information, user defined metadata can be stored on the TSM server of size up to 255 bytes. The user defined space is
employed to store Lustre information as well as a CRC32 check-sum and a magic id (see Fig. below).
![](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/object.meta.data.png)

# Copytool
As described above, a *copytool* is composed of a HSM proxy and data mover that runs on the same node. Loosely speaking, the
copytool receives archive, retrieve and delete actions from MDT node and triggers data moving operations. In more detail it is a
daemon process that requires a Lustre mount point for accessing Lustre files and processes actions *HSMA_ARCHIVE*, *HSMA_RESTORE*,
*HSMA_REMOVE* and *HSMA_CANCEL* as depicted below.

![](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/source.core.overview.png)

The functions *ct_archive(session)*, *ct_restore(session)* and *ct_remove(session)* implement the core actions for accessing named objects and for transferring data between the Lustre mount point and the TSM storage.

# Implementation Details
For achieving high data throughput by means of parallelism the copytool employs the producer-consumer model. The data
structure of the model is a concurrent queue. The producer is a single thread receiving HSM action items from the MDS’s and
enqueues the items in the queue. There are multiple consumer threads, each having a session opened to the TSM server, which are
dequeueing items and executing the HSM actions (see Fig. below).
![](https://raw.githubusercontent.com/tstibor/ltsm.github.io/master/doc/images/produce.consumer.model.png)

Overview of a single-producer multiple-consumer model for a produce thread (P-Thread) and four consumer threads
(C-Threads). Three C-Threads dequeued HSM action items and are executing the actions. The P-Thread received from the MDS
a new HSM action item and enqueues the item into the concurrent queue. Once the enqueue operation is successful, C-Thread_4
will be notified to dequeue and execute new HSM action.

# Getting Started <a id="getting.started"></a>
Before using LTSM a working access to a TSM server is required. One can install for testing purposes a fully working
TSM server (e.g. inside a KVM) for a period of 30 days before the license expires. A complete installation and setup guide is provided
at [TSM Server Installation Guide](https://github.com/tstibor/ltsm.github.io/tree/master/doc/tutorial).
Download and install the TSM API client provided at [7.1.X.Y-TIV-TSMBAC-LinuxX86_DEB.tar](http://ftp.software.ibm.com/storage/tivoli-storage-management/maintenance/client/v7r1/Linux/LinuxX86_DEB/BA/).
Make sure to install `tivsm-api64.amd64.deb` where the header files `dapitype.h, dsmapips.h, ...` and the low-level library `libApiTSM64.so` are provided.

### Compile *lhsmtool_tsm* and *ltsmc*

Clone the git repository
```
git clone https://github.com/tstibor/ltsm
```
and execute
```
./autogen.sh && ./configure CFLAGS='-g -DDEBUG -O0' --enable-tests
```
If Lustre header files are not found you will get the message
```
...
configure: WARNING: cannot find Lustre header files, use --with-lustre-headers=PATH
configure: WARNING: cannot find Lustre API headers and/or Lustre API library
...
configured ltsm:

build tsmapi library and ltsmc : yes
build lhsmtool_tsm             : no
build test suite               : yes
```
and the console client *ltsmc* is built only. For building also the Lustre Copytool thus make sure the Lustre header files
and the Lustre library `liblustreapi.so` are available and the paths are correctly specified e.g.
```
./autogen.sh && ./configure CFLAGS='-g -DDEBUG -O0' --with-lustre-src=/usr/local/include/lustre LDFLAGS='-L/usr/local/lib' --with-tsm-headers=/opt/tivoli/tsm/client/api/bin64/sample --enable-tests
```

If *required* TSM and *optional* Lustre header files and libraries are found the following executable files are provided:
  * `src/lhsmtool_tsm` (Lustre TSM Copytool)
  * `src/ltsmc` (Console client)
  * `src/test/test_cds` (Test suite for data structures (linked-list, queue, hashtable, etc.))
  * `src/test/test_tsmapi` (Test suite for *tsmapi*)
  * `src/test/ltsmbench` (Benchmark suite for measuring threaded archive/retrieve performance)

### Install or Build DEB/RPM Package

Download and install already built Debian Stretch package [ltsm_0.8.0_amd64.deb](https://github.com/tstibor/ltsm.github.io/tree/master/packages/deb) or CentOS 7.7 package [ltsm-0.8.0-1.x86_64.rpm](https://github.com/tstibor/ltsm.github.io/tree/master/packages/rpm). In addition you can build the rpm package
yourself as follows
```
git clone https://github.com/tstibor/ltsm && cd ltsm && ./autogen.sh && ./configure && make rpms
```
or the deb package
```
git clone https://github.com/tstibor/ltsm && cd ltsm && ./autogen.sh && ./configure && make debs
```
Make sure your Linux distribution supports [systemd](https://freedesktop.org/wiki/Software/systemd/).
For customizing the options of systemd started `lhsmtool_tsm` daemon, edit file `/etc/default/lhsmtool_tsm`.

## Lustre Copytool
Make sure you have enabled the HSM functionality on the MGS/MDS, e.g. `lctl set_param 'mdt.<LNAME>-MDT0000.hsm_control=enabled`
and TSM server is running as well as `dsm.sys` contains the proper entries such as
```
SErvername              polaris-kvm-tsm-server

debian/changelog
ltsm (0.9.1) unstable; urgency=low

  * Finalize remove of FSQ project.

 -- Thomas Stibor <t.stibor@gsi.de>  Mon, 13 Feb 2023 11:22:40 +0100

ltsm (0.9.0) unstable; urgency=low

  * Remove FSQ to keep LTSM as a Lustre TSM copytool project.

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 21 Jul 2022 11:08:50 +0200

ltsm (0.8.3-47) unstable; urgency=low

  * Automated changelog entry update

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 21 Jul 2022 11:08:50 +0200

ltsm (0.8.4) unstable; urgency=medium

  * Implement FSQ protocol version 1

 -- Thomas Stibor <thomas@stibor.net>  Mon, 30 May 2022 09:56:55 +0100

ltsm (0.8.3-2) unstable; urgency=medium

  * Improvement: Remove useless mutex lock on llapi_hsm_request

 -- Thomas Stibor <thomas@stibor.net>  Tue, 26 Jan 2021 09:21:55 +0100

ltsm (0.8.3) unstable; urgency=medium

  * Fix: llapi_hsm_action_progress: Cannot allocate memory (12)

 -- Thomas Stibor <thomas@stibor.net>  Thu, 14 Jan 2021 10:22:14 +0100

ltsm (0.8.2) unstable; urgency=low

  * Autoconf requires Lustre header files not full Lustre sources.
  * Add script ltsmsync.sh into package.

 -- Thomas Stibor <t.stibor@gsi.de>  Fri, 24 Nov 2020 11:06:51 +0100

ltsm (0.8.1) unstable; urgency=medium

  * TSM file system daemon bug fixes.

 -- Thomas Stibor <t.stibor@gsi.de>  Fri, 30 Jan 2020 12:40:30 +0200

ltsm (0.8.0) unstable; urgency=medium

  * Added Lustre TSM file system daemon.

 -- Thomas Stibor <t.stibor@gsi.de>  Wed, 15 Jan 2020 13:25:30 +0200

ltsm (0.7.3) unstable; urgency=medium

  * Lower and upper date/time bound for querying.
  * Display in query results the CRC32 value.

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 09 Aug 2018 10:53:30 +0200

ltsm (0.7.2) unstable; urgency=medium

  * Introduce conf file option setting and man page update/corrections.

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 05 Apr 2018 13:40:49 +0200

ltsm (0.7.1) unstable; urgency=medium

  * Update systemd requirements and fix archive_id handling.

 -- Thomas Stibor <t.stibor@gsi.de>  Mon, 11 Dec 2017 12:26:50 +0200

ltsm (0.7.0) unstable; urgency=medium

  * Build from official master upstream.

 -- Thomas Stibor <t.stibor@gsi.de>  Wed, 16 Nov 2017 12:42:32 +0200

ltsm (0.6.0-1) unstable; urgency=medium

  * Build from official master upstream.

 -- Thomas Stibor <t.stibor@gsi.de>  Wed, 08 Nov 2017 15:54:32 +0200

ltsm (0.5.7-1) unstable; urgency=medium

  * Build from official master upstream.

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 28 Sep 2017 16:51:32 +0200

ltsm (0.5.6-1) unstable; urgency=low

  * Initial Release.

 -- Thomas Stibor <t.stibor@gsi.de>  Thu, 28 Sep 2017 11:49:14 +0200


src/ltsmc.c
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 only,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License version 2 for more details (a copy is included
 * in the LICENSE file that accompanied this code).
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Copyright (c) 2016, 2017, GSI Helmholtz Centre for Heavy Ion Research
 */

#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <getopt.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <zlib.h>
#include "ltsmapi.h"
#include "measurement.h"

MSRT_DECLARE(tsm_archive_fpath);
MSRT_DECLARE(tsm_retrieve_fpath);

struct options {
	int o_archive;
	int o_retrieve;
	int o_query;
	int o_delete;
	int o_pipe;
	int o_verbose;
	int o_latest;
	int o_recursive;
	int o_checksum;
	int o_sort;
	char o_servername[DSM_MAX_SERVERNAME_LENGTH + 1];
	char o_node[DSM_MAX_NODE_LENGTH + 1];
	char o_owner[DSM_MAX_OWNER_LENGTH + 1];
	char o_password[DSM_MAX_VERIFIER_LENGTH + 1];
	char o_fsname[DSM_MAX_FSNAME_LENGTH + 1];
	char o_fstype[DSM_MAX_FSTYPE_LENGTH + 1];
	char o_desc[DSM_MAX_DESCR_LENGTH + 1];
	char o_conf[MAX_OPTIONS_LENGTH + 1];
	dsmDate o_date_lower_bound;
	dsmDate o_date_upper_bound;
};

struct options opt = {
	.o_verbose	    = API_MSG_NORMAL,
	.o_servername	    = {0},
	.o_node		    = {0},
	.o_owner	    = {0},
	.o_password	    = {0},
	.o_fsname	    = {0},
	.o_fstype	    = {0},
	.o_desc		    = {0},
	.o_conf		    = {0},
	.o_date_lower_bound = {DATE_MINUS_INFINITE, 1, 1, 0, 0, 0},
	.o_date_upper_bound = {DATE_PLUS_INFINITE, 12, 31, 23, 59, 59},
	.o_checksum	    = 0,
	.o_sort		    = SORT_NONE
};

static void usage(const char *cmd_name, const int rc)
{
	dsmApiVersionEx libapi_ver = get_libapi_ver();
	dsmAppVersion appapi_ver = get_appapi_ver();

	fprintf(stdout, "usage: %s [options] <files|directories|wildcards>\n"
		"\t--archive\n"
		"\t--retrieve\n"
		"\t--query\n"
		"\t--delete\n"
		"\t--pipe\n"
		"\t--checksum\n"
		"\t-l, --latest [retrieve object with latest timestamp when multiple exists]\n"
		"\t-x, --prefix [retrieve prefix directory]\n"
		"\t-r, --recursive [archive directory and all sub-directories]\n"
		"\t-t, --sort={ascending, descending, restore} [sort query in date or restore order]\n"
		"\t-f, --fsname <string> [default: '/']\n"
		"\t-d, --description <string>\n"
		"\t-n, --node <string>\n"
		"\t-o, --owner <string>\n"
		"\t-p, --password <string>\n"
		"\t-s, --servername <string>\n"
		"\t-v, --verbose {error, warn, message, info, debug} [default: %s]\n"
		"\t-c, --conf <file>\n"
		"\t-y, --datelow <string>\n"
		"\t-z, --datehigh <string>\n"
		"\t-h, --help\n"
		"\nIBM API library version: %d.%d.%d.%d, "
		"IBM API application client version: %d.%d.%d.%d\n"
		"version: %s © 2017 by GSI Helmholtz Centre for Heavy Ion Research\n",
		cmd_name,
		LOG_LEVEL_HUMAN_STR(opt.o_verbose),
		libapi_ver.version, libapi_ver.release, libapi_ver.level,
		libapi_ver.subLevel,
		appapi_ver.applicationVersion, appapi_ver.applicationRelease,
		appapi_ver.applicationLevel, appapi_ver.applicationSubLevel,
		PACKAGE_VERSION);
	exit(rc);
}

static int is_valid(char *arg, const int lower_bound, const int upper_bound)
{
	char *end = NULL;
	int val = strtol(arg, &end, 10);

	if (*end != '\0') {
		CT_ERROR(-EINVAL, "invalid argument: '%s'", arg);
		return -EINVAL;
	}
	if (!(val >= lower_bound && val <= upper_bound)) {
		CT_ERROR(-EINVAL, "invalid argument: '%s'", arg);
		return -EINVAL;
	}

	return val;
}

static int parse_date_time(char *arg, dsmDate *dsm_date)
{
	const char *delim = ":";
	char *token;
	char *saveptr;
	uint16_t cnt = 0;
	int val;

	/* YYYY:MM:DD:hh:mm:ss, e.g. 2018:05:28:23:15:59 */
	token = strtok_r(arg, delim, &saveptr);
	while (token) {
		if (cnt == 0) { /* Year, 16-bit integer (e.g., 1990). */
			val = is_valid(token, 0, 0xFFFF);
			if (val < 0)
				return val;
			dsm_date->year = val;
		}
		else if (cnt == 1) { /* Month, 8-bit integer (1 - 12). */
			val = is_valid(token, 1, 12);
			if (val < 0)
				return val;
			dsm_date->month = val;
		}
		else if (cnt == 2) { /* Day, 8-bit integer (1 - 31). */
			val = is_valid(token, 1, 31);
			if (val < 0)
				return val;
			dsm_date->day = val;
		}
		else if (cnt == 3) { /* Hour, 8-bit integer (0 - 23). */
			val = is_valid(token, 0, 23);
			if (val < 0)
				return val;
			dsm_date->hour = val;
		}
		else if (cnt == 4) { /* Minute, 8-bit integer (0 - 59). */
			val = is_valid(token, 0, 59);
			if (val < 0)
				return val;
			dsm_date->minute = val;
		}
		else if (cnt == 5) { /* Second, 8-bit integer (0 - 59). */
			val = is_valid(token, 0, 59);
			if (val < 0)
				return val;
			dsm_date->second = val;
		}
		else {
			CT_ERROR(-EINVAL, "invalid argument: '%s'", token);
			return -EINVAL;
		}
		cnt++;
		token = strtok_r(NULL, delim, &saveptr);
	}
	return (cnt <= 6 ? 0 : -EINVAL);
}

static void read_conf(const char *filename)
{
	int rc;
	struct kv_opt kv_opt = {
		.N	     = 0,
		.kv	     = NULL
	};

	rc = parse_conf(filename, &kv_opt);
	if (!rc) {
		for (uint8_t n = 0; n < kv_opt.N; n++) {
			if (OPTNCMP("servername", kv_opt.kv[n].key))
				strncpy(opt.o_servername, kv_opt.kv[n].val,
					sizeof(opt.o_servername));
			else if (OPTNCMP("node", kv_opt.kv[n].key))
				strncpy(opt.o_node, kv_opt.kv[n].val,
					sizeof(opt.o_node));
			else if (OPTNCMP("owner", kv_opt.kv[n].key))
				strncpy(opt.o_owner, kv_opt.kv[n].val,
					sizeof(opt.o_owner));
			else if (OPTNCMP("password", kv_opt.kv[n].key))
				strncpy(opt.o_password, kv_opt.kv[n].val,
					sizeof(opt.o_password));
			else if (OPTNCMP("fsname", kv_opt.kv[n].key))
				strncpy(opt.o_fsname, kv_opt.kv[n].val,
					sizeof(opt.o_fsname));
			else if (OPTNCMP("verbose", kv_opt.kv[n].key)) {
				rc = parse_verbose(kv_opt.kv[n].val,
						   &opt.o_verbose);
				if (rc)
					CT_WARN("wrong value '%s' for option '%s'"
						" in conf file '%s'",
						kv_opt.kv[n].val, kv_opt.kv[n].key,
						filename);
			} else
				CT_WARN("unknown option value '%s %s' in conf"
					" file '%s'", kv_opt.kv[n].key,
					kv_opt.kv[n].val, filename);
		}
	}

	if (kv_opt.kv) {
		free(kv_opt.kv);
		kv_opt.kv = NULL;
		kv_opt.N = 0;
	}
}

static void sanity_arg_check(const char *argv)
{
	uint8_t count = 0;
	count = opt.o_archive  == 1 ? count + 1 : count;
	count = opt.o_retrieve == 1 ? count + 1 : count;
	count = opt.o_delete   == 1 ? count + 1 : count;
	count = opt.o_query    == 1 ? count + 1 : count;
	count = opt.o_checksum == 1 ? count + 1 : count;
	count = opt.o_pipe     == 1 ? count + 1 : count;

	if (count == 0) {
		CT_ERROR(0, "missing argument --archive, --retrieve,"
			 " --query, --delete, --pipe or --checksum");
		usage(argv, 1);
	} else if (count != 1) {
		CT_ERROR(0, "multiple incompatible arguments"
			 " --archive, --retrieve, --query, --delete, --pipe or "
			 "--checksum");
		usage(argv, 1);
	}

	/* There are no additional required arguments when
	   --checksum is chosen. */
	if (opt.o_checksum)
		return;

	/* Required arguments. */
	if (!opt.o_node[0]) {
		CT_ERROR(0, "missing argument -n, --node <string>");
		usage(argv, 1);
	} else if (!opt.o_servername[0]) {
		CT_ERROR(0, "missing argument -s, --servername <string>");
		usage(argv, 1);
	} else if (!opt.o_fsname[0])
		strncpy(opt.o_fsname, DEFAULT_FSNAME, DSM_MAX_FSNAME_LENGTH);
}

static int parseopts(int argc, char *argv[])
{
	struct option long_opts[] = {
		{.name = "archive",	.has_arg = no_argument,       .flag = &opt.o_archive,  .val = 1},
		{.name = "retrieve",	.has_arg = no_argument,       .flag = &opt.o_retrieve, .val = 1},
		{.name = "query",	.has_arg = no_argument,       .flag = &opt.o_query,    .val = 1},
		{.name = "delete",	.has_arg = no_argument,       .flag = &opt.o_delete,   .val = 1},
		{.name = "pipe",        .has_arg = no_argument,       .flag = &opt.o_pipe,     .val = 1},
		{.name = "checksum",    .has_arg = no_argument,       .flag = &opt.o_checksum, .val = 1},
		{.name = "latest",	.has_arg = no_argument,       .flag = NULL,            .val = 'l'},
		{.name = "recursive",	.has_arg = no_argument,       .flag = NULL,	       .val = 'r'},
		{.name = "sort",	.has_arg = required_argument, .flag = NULL,	       .val = 't'},
		{.name = "fsname",	.has_arg = required_argument, .flag = NULL,	       .val = 'f'},
		{.name = "description", .has_arg = required_argument, .flag = NULL,	       .val = 'd'},
		{.name = "node",	.has_arg = required_argument, .flag = NULL,	       .val = 'n'},
		{.name = "owner",	.has_arg = required_argument, .flag = NULL,	       .val = 'o'},
		{.name = "password",	.has_arg = required_argument, .flag = NULL,	       .val = 'p'},
		{.name = "servername",	.has_arg = required_argument, .flag = NULL,	       .val = 's'},
		{.name = "verbose",	.has_arg = required_argument, .flag = NULL,	       .val = 'v'},
		{.name = "prefix",	.has_arg = required_argument, .flag = NULL,	       .val = 'x'},
		{.name = "conf",	.has_arg = required_argument, .flag = NULL,	       .val = 'c'},
		{.name = "datelow",	.has_arg = required_argument, .flag = NULL,	       .val = 'y'},
		{.name = "datehigh",	.has_arg = required_argument, .flag = NULL,	       .val = 'z'},
		{.name = "help",	.has_arg = no_argument,       .flag = NULL,	       .val = 'h'},
		{.name = NULL}
	};

	int c;
	while ((c = getopt_long(argc, argv, "lrt:f:d:n:o:p:s:v:x:c:y:z:h",
				long_opts, NULL)) != -1) {
		switch (c) {
		case 'l': {
			opt.o_latest = 1;
			break;
		}
		case 'r': {
			opt.o_recursive = 1;
			set_recursive(bTrue);
			break;
		}
		case 't': {
			if (OPTNCMP("none", optarg))
				opt.o_sort = SORT_NONE;
			else if (OPTNCMP("ascending", optarg))
				opt.o_sort = SORT_DATE_ASCENDING;
			else if (OPTNCMP("descending", optarg))
				opt.o_sort = SORT_DATE_DESCENDING;
			else if (OPTNCMP("restore", optarg))
				opt.o_sort = SORT_RESTORE_ORDER;
			else {
				CT_ERROR(0, "wrong argument for -t, "
					"--sort '%s", optarg);
				usage(argv[0], 1);
			}
			break;
		}
		case 'f': {
			strncpy(opt.o_fsname, optarg, DSM_MAX_FSNAME_LENGTH);
			break;
		}
		case 'd': {
			strncpy(opt.o_desc, optarg, DSM_MAX_DESCR_LENGTH);
			break;
		}
		case 'n': {
			strncpy(opt.o_node, optarg, DSM_MAX_NODE_LENGTH);
			break;
		}
		case 'o': {
			strncpy(opt.o_owner, optarg, DSM_MAX_OWNER_LENGTH);
			break;
		}
		case 'p': {
			strncpy(opt.o_password, optarg,
				DSM_MAX_VERIFIER_LENGTH);
			break;
		}
		case 's': {
			strncpy(opt.o_servername, optarg,
				DSM_MAX_SERVERNAME_LENGTH);
			break;
		}
		case 'v': {
			int rc = parse_verbose(optarg, &opt.o_verbose);
			if (rc) {
				CT_ERROR(0, "wrong argument for -v, "
					 "--verbose '%s'", optarg);
				usage(argv[0], 1);
			}
			break;
		}
		case 'x': {
			set_prefix(optarg);
			break;
		}
		case 'c': {
			read_conf(optarg);
			break;
		}
		case 'y': {
			int rc = parse_date_time(optarg, &opt.o_date_lower_bound);
			if (rc) {
				CT_ERROR(0, "wrong argument for -y, "
					 "--datelow '%s'", optarg);
				usage(argv[0], 1);
			}
			break;
		}
		case 'z': {
			int rc = parse_date_time(optarg, &opt.o_date_upper_bound);
			if (rc) {
				CT_ERROR(0, "wrong argument for -z, "
					 "--datehigh '%s'", optarg);
				usage(argv[0], 1);
			}
			break;
		}
		case 'h': {
			usage(argv[0], 0);
			break;
		}
		case 0: {
			break;
		}
		default:
			return -EINVAL;
		}
	}

	sanity_arg_check(argv[0]);

	return 0;
}

static int progress_callback(struct progress_size_t *pg_size,
			      struct session_t *session)
{
	MSRT_DATA(tsm_archive_fpath, pg_size->cur);
	MSRT_DATA(tsm_retrieve_fpath, pg_size->cur);

	return 0;
}

int main(int argc, char *argv[])
{
	int rc;
	api_msg_set_level(opt.o_verbose);
	rc = parseopts(argc, argv);
	if (rc) {
		CT_WARN("try '%s --help' for more information", argv[0]);
		return 1;
	}

	size_t num_files_dirs = 0;
	char **files_dirs_arg = NULL;

	/* file or directory parameter */
	if (optind < argc) {
		size_t i = 0;
		num_files_dirs = argc - optind;
		files_dirs_arg = malloc(sizeof(char *) * num_files_dirs);
		while (optind < argc) {
			size_t len = strlen(argv[optind]);
			files_dirs_arg[i] = calloc(len + 1, sizeof(char));
			(files_dirs_arg[i])[len] = '\0';
			strncpy(files_dirs_arg[i++], argv[optind++], len);
		}
	}

	if (opt.o_checksum) {

		if (num_files_dirs == 0) {
			CT_ERROR(0, "missing argument <files>");
			usage(argv[0], 1);
		}

		uint32_t crc32sum = 0;
		for (size_t i = 0; i < num_files_dirs &&
			     files_dirs_arg[i]; i++) {
			rc = crc32file(files_dirs_arg[i], &crc32sum);
			if (rc)
				CT_WARN("calculation of crc32 for '%s' failed",
					files_dirs_arg[i]);
			else
				fprintf(stdout, "crc32: "
					"0x%08x (%010u), file: '%s'\n", crc32sum,
					crc32sum, files_dirs_arg[i]);
		}
		goto cleanup;
	}

	struct login_t login;
	login_init(&login, opt.o_servername,
		   opt.o_node, opt.o_password,
		   opt.o_owner, LINUX_PLATFORM,
		   opt.o_fsname, DEFAULT_FSTYPE);

	struct session_t session;
	memset(&session, 0, sizeof(session));

	session.qtable.multiple = opt.o_latest == 1 ? bFalse : bTrue;
	session.qtable.sort_by = opt.o_sort;

	rc = tsm_init(DSM_SINGLETHREAD);
	if (rc)
		goto cleanup;

	if (opt.o_pipe) {
		if (num_files_dirs == 0) {
			CT_ERROR(0, "missing argument <files>");
			usage(argv[0], 1);
		} else if (num_files_dirs > 1) {
			CT_ERROR(0, "too many arguments <files>");
			usage(argv[0], 1);
		}

		rc = tsm_fconnect(&login, &session);
		if (rc)
			goto cleanup_tsm;

		rc = tsm_fopen(opt.o_fsname, files_dirs_arg[0], opt.o_desc,
			       &session);
		if (rc) {
			goto cleanup_tsm;
		}

		char buf[TSM_BUF_LENGTH] = {0};
		size_t size;
		do {
			size = fread(buf, 1, TSM_BUF_LENGTH, stdin);
			if (ferror(stdin)) {
				session.tsm_file->err = EIO;
				CT_ERROR(EIO, "fread failed");
				break;
			}
			if (size == 0)
				break;
			rc = tsm_fwrite(buf, 1, size, &session);
			if (rc < 0) {
				CT_ERROR(errno, "tsm_fwrite failed");
				break;
			}
		} while (!feof(stdin));

		rc = tsm_fclose(&session);
		if (rc)
			CT_ERROR(errno, "tsm_fclose failed");

		goto cleanup_tsm;
	}

	session.progress = progress_callback;

	rc = tsm_connect(&login, &session);
	if (rc)
		goto cleanup_tsm;

	rc = tsm_query_session(&session);
	if (rc)
		goto cleanup_tsm;

	/* Handle operations on files and directories resp. */
	for (size_t i = 0; i < num_files_dirs &&
		     files_dirs_arg[i]; i++) {
		if (opt.o_query)
			rc = tsm_query_fpath(opt.o_fsname, files_dirs_arg[i],
					     opt.o_desc, &opt.o_date_lower_bound,
					     &opt.o_date_upper_bound, &session);
		else if (opt.o_retrieve) {
			MSRT_START(tsm_retrieve_fpath);
			rc = tsm_retrieve_fpath(opt.o_fsname, files_dirs_arg[i],
						opt.o_desc, -1, &session);
			MSRT_STOP(tsm_retrieve_fpath);
			MSRT_DISPLAY_RESULT(tsm_retrieve_fpath);
		}
		else if (opt.o_delete)
			rc = tsm_delete_fpath(opt.o_fsname, files_dirs_arg[i],
					      &session);
		else if (opt.o_archive) {
			MSRT_START(tsm_archive_fpath);
			rc = tsm_archive_fpath(opt.o_fsname,
					       files_dirs_arg[i],
					       opt.o_desc, -1, NULL, &session);
			MSRT_STOP(tsm_archive_fpath);
			MSRT_DISPLAY_RESULT(tsm_archive_fpath);
		}
		if (rc)
			goto cleanup_tsm;
	}

cleanup_tsm:
	tsm_disconnect(&session);
	tsm_cleanup(DSM_SINGLETHREAD);

cleanup:
	if (files_dirs_arg) {
		for (size_t i = 0; i < num_files_dirs; i++) {
			if (files_dirs_arg[i])
				free(files_dirs_arg[i]);
		}
		free(files_dirs_arg);
	}

	return rc;
}
